### PR TITLE
Adds hash-ref-key primitive

### PR DESCRIPTION
**Pull request by 97jaz**
For now, the operation is implemented on mutable tables in Chez
using a combination of hashtable-contains? and hashtable-cell.
A more efficient version will require modifying Chez.

See #2607

### CODE REVIEW

File: pkgs/racket-doc/scribblings/reference/hashes.scrbl [hunk 1]
```diff
@@ -273,6 +273,75 @@ result:
...
+
+(define table (make-hash))
+(hash-set! table original-key 'one)
+(hash-set! table key-copy 'two)
+
+(eq? (hash-ref-key table "hello") original-key)
+(eq? (hash-ref-key table "hello") key-copy)
+]
+
+Conversely, an immutable hash retains the key that was most-recently
```
Comment (97jaz) [inline pkgs/racket-doc/scribblings/reference/hashes.scrbl]: This behavior could be easily changed to match that of immutable hashes, if we want to do that.

File: racket/src/cs/rumble/hash.ss [hunk 2]
```diff
@@ -1065,6 +1116,17 @@
                             hash-procs-ref
                             ht k none))
 
+(define (impersonate-hash-ref-key ht k)
+  (impersonate-hash-ref/set 'hash-ref-key #f
+                            (lambda (ht k v) (hash-ref-key ht k none))
+                            (lambda (procs ht k none-v)
+                              (let-values ([(new-k _) (|#%app| (hash-procs-ref procs) ht k)])
+                                (values new-k
+                                        (lambda (ht k none-v)
```
Comment (97jaz) [inline racket/src/cs/rumble/hash.ss]: This lambda requires a closure  (as `procs` is bound by the enclosing lambda). Not sure if that's okay.

Also, if `ht` is a hash chaperone and the resulting key isn't `chaperone-of?` the original, the error message will describe the key as a "value":
```racket
                   [else
                    (let ([new-r (new-v-or-wrap next-ht new-k r)])
                      (when chaperone?
                        (unless (chaperone-of? new-r r)
                          (raise-chaperone-error who "value" new-r r)))
                      new-r)]
```

To correct that, maybe we should add a `what-r` parameter to `impersonate-hash-ref/set`?

File: pkgs/racket-doc/scribblings/reference/hashes.scrbl [hunk 3]
```diff
@@ -273,6 +273,75 @@ result:
...
+
+(eq? (hash-ref-key table2 "hello") original-key)
+(eq? (hash-ref-key table2 "hello") key-copy)
+]
+
+If @racket[hash] is an @tech{impersonator}, then the returned key
+will be determined as described in the documentation to
+@racket[impersonate-hash].
+
+@see-also-caveats[]}
```
Comment (mflatt) [inline pkgs/racket-doc/scribblings/reference/hashes.scrbl]: Need a `history` annotation here.

File: racket/src/cs/rumble/hash.ss (hunk 2 above)
Comment (mflatt) [reply racket/src/cs/rumble/hash.ss]: Adding a `what-r` to `impersonate-hash-ref/set` sounds like a good idea.

File: racket/src/cs/rumble/intmap.ss [hunk 4]
```diff
@@ -63,35 +61,42 @@
...
+          (loop (Br-left t))
+          (loop (Br-right t)))]
 
-   [(Lf? t)
-    (if (key=? et key (Lf-key t))
-        (Lf-value t)
-        ($fail def))]
+     [(Lf? t)
+      (if (key=? et key (Lf-key t))
+          (with-leaf t)
```
Comment (mflatt) [inline racket/src/cs/rumble/intmap.ss]: Does replacing the known selector used with a call to an unknown function affect performance? I would not expect `do-$intmap-ref` to be inlined, but I may guess wrong.

File: racket/src/racket/src/schvers.h [hunk 5]
```diff
@@ -16,7 +16,7 @@
 #define MZSCHEME_VERSION_X 7
 #define MZSCHEME_VERSION_Y 4
 #define MZSCHEME_VERSION_Z 0
-#define MZSCHEME_VERSION_W 1
+#define MZSCHEME_VERSION_W 3
```
Comment (mflatt) [inline racket/src/racket/src/schvers.h]: Also need to update "info.rkt" in the "base" package.

Comment (97jaz) [reply racket/src/racket/src/schvers.h]: Fixed in 2faf3b6cfc5d7d14fa2b3b26b39289c7e6b1a59a

File: pkgs/racket-doc/scribblings/reference/hashes.scrbl (hunk 3 above)
Comment (97jaz) [reply pkgs/racket-doc/scribblings/reference/hashes.scrbl]: Fixed in ea8574a786565cf0221862c05578aeb9ab5349ec

File: racket/src/cs/rumble/intmap.ss (hunk 4 above)
Comment (97jaz) [reply racket/src/cs/rumble/intmap.ss]: Performance seems minimally effected by this change. 

To give you some more context, originally I implemented `do-$intmap-ref` as syntax, after the model of [`lookup-keyval` in Chez's implementation of mutable eq-hashtables](https://github.com/cisco/ChezScheme/blob/58663d92a353969e3b27de2e1a44df571036e1b2/s/library.ss#L1380). However, this had a marked negative effect on the "large table" tests from `hash-demo`:

Original implementation:
```
large tables
cpu time: 123 real time: 123 gc time: 2   MB: 292
cpu time: 162 real time: 162 gc time: 2   MB: 292
```

With `do-$intmap-ref` as syntax:
```
large tables
cpu time: 162 real time: 162 gc time: 2   MB: 292
cpu time: 336 real time: 336 gc time: 2   MB: 292
```

With `do-$intmap-ref` as function:
```
large tables
cpu time: 125 real time: 125 gc time: 2   MB: 292
cpu time: 164 real time: 164 gc time: 3   MB: 292
```

The other tests in that demo, however, are unaffected. I have no idea what to make of this. If Chez does loop unrolling, then I could see how the increased code size might have some negative cache effects, but I don't think it does any unrolling.

File: racket/src/cs/rumble/hash.ss (hunk 2 above)
Comment (97jaz) [reply racket/src/cs/rumble/hash.ss]: Fixed in 42425bb70de068932e0aaced5d133fe7583f8b76

File: racket/src/cs/rumble/intmap.ss (hunk 4 above)
Comment (97jaz) [reply racket/src/cs/rumble/intmap.ss]: Well, I guess the [unrolling story](https://github.com/cisco/ChezScheme/blob/58663d92a353969e3b27de2e1a44df571036e1b2/s/cp0.ss#L79) is a bit more complicated than that.

**Comment by mflatt** [comment]
Thanks! The version number at master is now already 7.4.0.2, so if you could rebase, update to 7.4.0.3, and squash, that would make the cleanest merge. But I can also work with these commits if you prefer.

**Comment by 97jaz** [comment]
- version (and history annotation) updated to 7.4.0.3
- squashed
- rebased against upstream master

**Pull request merged**